packages/playwright/src/runner/watchMode.ts

```diff
@@ -15,130 +15,115 @@
-        let set = this._dirtyTestFiles.get(project);
-        if (!set) {
-          set = new Set();
-          this._dirtyTestFiles.set(project, set);
+interface WatchModeOptions {
+  files?: string[];
+  projects?: string[];
+  grep?: string;
+  onlyChanged?: string;
+}
+
+export async function runWatchModeLoop(configLocation: ConfigLocation, initialOptions: WatchModeOptions): Promise<FullResult['status']> {
+  const options: WatchModeOptions = { ...initialOptions };
+
+  const testServerDispatcher = new TestServerDispatcher(configLocation);
+  const inMemorySocket = new InMemoryServerSocket(
+      async data => {
+        const { id, method, params } = JSON.parse(data);
+        try {
+          const result = await testServerDispatcher.transport.dispatch(method, params);
+          inMemorySocket.emit('message', { data: JSON.stringify({ id, result }) });
+        } catch (e) {
+          inMemorySocket.emit('message', { data: JSON.stringify({ id, error: String(e) }) });
         }
-        filteredFiles.map(f => set!.add(f));
-        hasMatches = true;
       }
+  );
+  testServerDispatcher.transport.sendEvent = (method, params) => {
+    inMemorySocket.emit('message', { data: JSON.stringify({ method, params }) });
+  };
+  const testServerConnection = new TestServerConnection(inMemorySocket);
+  inMemorySocket.emit('open');
 
-      if (!hasMatches)
-        return;
+  const dirtyTestFiles: string[] = [];
+  const onDirtyTestFiles: { resolve?(): void } = {};
+  const failedTestIdCollector = new Set<string>();
+  const projectNames: string[] = [];
 
-      if (this._timer)
-        clearTimeout(this._timer);
-      this._timer = setTimeout(() => {
-        this._notifyDirtyFiles?.();
-      }, 250);
-    });
+  testServerConnection.onTestFilesChanged(async ({ testFiles: changedFiles }) => {
+    if (changedFiles.length === 0)
+      return;
 
-  }
+    const { report } = await testServerConnection.listTests({ locations: options.files, projects: options.projects, grep: options.grep, onlyChanged: options.onlyChanged });
+    const rootDir = report.find(r => r.method === 'onConfigure')!.params.config.rootDir;
+    for (const suite of report.find(r => r.method === 'onProject')!.params.project.suites) {
+      const suitePath = path.join(rootDir, suite.location.file);
+      if (changedFiles.includes(suitePath))
+        dirtyTestFiles.push(suitePath);
+    }
 
-  async onDirtyTestFiles(): Promise<void> {
-    if (this._dirtyTestFiles.size)
+    if (dirtyTestFiles.length === 0)
       return;
-    await new Promise<void>(f => this._notifyDirtyFiles = f);
-  }
 
-  takeDirtyTestFiles(): Map<FullProjectInternal, Set<string>> {
-    const result = this._dirtyTestFiles;
-    this._dirtyTestFiles = new Map();
-    return result;
-  }
-}
+    onDirtyTestFiles.resolve?.();
+  });
+  testServerConnection.onReport(report => {
+    if (report.method === 'onProject') {
+      const { name } = report.params.project;
+      projectNames.push(name);
```

Comment:
We seems to push into this array over and over again, on each test run, don't we?

Reply:
We do! We could either clear it on every new test run, or make it a `Set`. I've turned it into a `Set` in 05ea3955564ef4f46b9a2d3c6f100b465fc867e9.